oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/endpoint/NimbusJwtClientAuthenticationParametersConverter.java

```diff
@@ -0,0 +1,248 @@
+		}
+
+		JoseHeader.Builder headersBuilder = JoseHeader.withAlgorithm(jwsAlgorithm);
+
+		Instant issuedAt = Instant.now();
+		Instant expiresAt = issuedAt.plus(Duration.ofSeconds(60));
+
+		// @formatter:off
+		JwtClaimsSet.Builder claimsBuilder = JwtClaimsSet.builder()
+				.issuer(clientRegistration.getClientId())
+				.subject(clientRegistration.getClientId())
+				.audience(Collections.singletonList(clientRegistration.getProviderDetails().getTokenUri()))
+				.id(UUID.randomUUID().toString())
+				.issuedAt(issuedAt)
+				.expiresAt(expiresAt);
+		// @formatter:on
+
+		JwtClientAuthenticationContext<T> context = new JwtClientAuthenticationContext<>(authorizationGrantRequest,
+				headersBuilder.headers, claimsBuilder.claims);
+		this.jwtCustomizer.accept(context);
+
+		JoseHeader joseHeader = headersBuilder.build();
+		JwtClaimsSet jwtClaimsSet = claimsBuilder.build();
+
+		NimbusJwsEncoder jwsEncoder = this.jwsEncoders.computeIfAbsent(clientRegistration.getRegistrationId(),
+				(clientRegistrationId) -> {
+					JWKSource<SecurityContext> jwkSource = new ImmutableJWKSet<>(new JWKSet(jwk));
+					return new NimbusJwsEncoder(jwkSource);
+				});
+
+		Jwt jws = jwsEncoder.encode(joseHeader, jwtClaimsSet);
+
+		MultiValueMap<String, String> parameters = new LinkedMultiValueMap<>();
+		parameters.set(OAuth2ParameterNames.CLIENT_ASSERTION_TYPE, CLIENT_ASSERTION_TYPE_VALUE);
+		parameters.set(OAuth2ParameterNames.CLIENT_ASSERTION, jws.getTokenValue());
+
+		return parameters;
+	}
+
+	/**
+	 * Sets the {@link Consumer} to be provided the opportunity to customize the
+	 * {@link Jwt} headers and/or claims.
+	 * @param jwtCustomizer the {@link Consumer} to be provided the opportunity to
+	 * customize the {@link Jwt} headers and/or claims
+	 */
+	public void setJwtCustomizer(Consumer<JwtClientAuthenticationContext<T>> jwtCustomizer) {
+		Assert.notNull(jwtCustomizer, "jwtCustomizer cannot be null");
+		this.jwtCustomizer = jwtCustomizer;
+	}
+
+	private static JwsAlgorithm resolveAlgorithm(JWK jwk) {
+		JwsAlgorithm jwsAlgorithm = null;
+
+		if (jwk.getAlgorithm() != null) {
+			jwsAlgorithm = SignatureAlgorithm.from(jwk.getAlgorithm().getName());
+			if (jwsAlgorithm == null) {
+				jwsAlgorithm = MacAlgorithm.from(jwk.getAlgorithm().getName());
+			}
+		}
+
+		if (jwsAlgorithm == null) {
+			if (KeyType.RSA.equals(jwk.getKeyType())) {
+				jwsAlgorithm = SignatureAlgorithm.RS256;
+			}
+			else if (KeyType.EC.equals(jwk.getKeyType())) {
+				jwsAlgorithm = SignatureAlgorithm.ES256;
+			}
+			else if (KeyType.OCT.equals(jwk.getKeyType())) {
+				jwsAlgorithm = MacAlgorithm.HS256;
+			}
+		}
+
+		return jwsAlgorithm;
+	}
+
+	/**
+	 * A context that provides access to the {@link Jwt} headers and/or claims allowing
+	 * for customization.
+	 *
+	 * @param <T> the type of {@link AbstractOAuth2AuthorizationGrantRequest}
+	 */
+	public static final class JwtClientAuthenticationContext<T extends AbstractOAuth2AuthorizationGrantRequest> {
```

Comment:
Is there a reason you went with creating `JwtClientAuthenticationContext` vs two converters like was done with the headers and claims in `AbstractOAuth2AuthorizationGrantRequestEntityConverter`

Reply:
See [comment](https://github.com/spring-projects/spring-security/pull/9520#discussion_r609660460)